File: ring-android/app/src/main/java/cx/ring/account/DeviceAdapter.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package cx.ring.account;

import android.content.Context;
import android.graphics.Typeface;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Map;

import cx.ring.R;
import cx.ring.views.TwoButtonEditText;

public class DeviceAdapter extends BaseAdapter {
    private final Context mContext;
    private final ArrayList<Map.Entry<String, String>> mDevices = new ArrayList<>();

    private String mCurrentDeviceId;
    private DeviceRevocationListener mListener;

    public DeviceAdapter(Context c, Map<String, String> devices, String currentDeviceId,
                         DeviceRevocationListener listener) {
        mContext = c;
        setData(devices, currentDeviceId);
        mListener = listener;
    }

    public void setData(Map<String, String> devices, String currentDeviceId) {
        mDevices.clear();
        mCurrentDeviceId = currentDeviceId;
        if (devices != null && !devices.isEmpty()) {
            mDevices.ensureCapacity(devices.size());
            mDevices.addAll(devices.entrySet());
        }
        for (int i = 0; i < mDevices.size(); i++) {
            if(mDevices.get(i).getKey().contentEquals(mCurrentDeviceId)) {
                mDevices.remove(mDevices.get(i));
            }
        }
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return mDevices.size();
    }

    @Override
    public Object getItem(int i) {
        return mDevices.get(i);
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(final int i, View view, ViewGroup parent) {
        if (view == null) {
            view = LayoutInflater.from(mContext).inflate(R.layout.item_device, parent, false);
        }
        boolean isCurrentDevice = mDevices.get(i).getKey().contentEquals(mCurrentDeviceId);

        TwoButtonEditText devId = view.findViewById(R.id.txt_device_id);
        TextView thisDevice = view.findViewById(R.id.txt_device_thisflag);
        devId.setText(mDevices.get(i).getValue());
        String hint = mDevices.get(i).getKey();
        hint = hint.substring(0, (int) (hint.length() * 0.66));
        devId.setHint(hint);

        if (isCurrentDevice) {
            thisDevice.setVisibility(View.VISIBLE);
            devId.setLeftDrawable(R.drawable.baseline_edit_twoton_24dp);
            devId.setLeftDrawableOnClickListener(view1 -> {
                if (mListener != null) {
                    mListener.onDeviceRename();
                }
            });
        } else {
            thisDevice.setVisibility(View.GONE);
            devId.setLeftDrawable(R.drawable.baseline_cancel_24);
            devId.setLeftDrawableOnClickListener(view12 -> {
                if (mListener != null) {
                    mListener.onDeviceRevocationAsked(mDevices.get(i).getKey());
                }
            });
        }

        return view;
    }

    public interface DeviceRevocationListener {
        void onDeviceRevocationAsked(String deviceId);

        void onDeviceRename();
    }
}

File: jami-android/app/src/main/java/cx/ring/account/DeviceAdapter.kt
/*
 *  Copyright (C) 2004-2023 Savoir-faire Linux Inc.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package cx.ring.account

import android.content.Context
import android.widget.BaseAdapter
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import cx.ring.R
import cx.ring.views.TwoButtonEditText
import android.widget.TextView
import java.util.ArrayList

class DeviceAdapter(
    private val mContext: Context, devices: Map<String, String>?,
    currentDeviceId: String?,
    private val mListener: DeviceRevocationListener
) : BaseAdapter() {
    private val mDevices = ArrayList<Map.Entry<String, String>>()
    private var mCurrentDeviceId: String? = null

    fun setData(devices: Map<String, String>?, currentDeviceId: String?) {
        mDevices.clear()
        mCurrentDeviceId = currentDeviceId
        if (devices != null && devices.isNotEmpty()) {
            mDevices.ensureCapacity(devices.size)
            mDevices.addAll(devices.entries)
        }
        mDevices.removeAll { d ->d.key == mCurrentDeviceId }
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return mDevices.size
    }

    override fun getItem(i: Int): Any {
        return mDevices[i]
    }

    override fun getItemId(i: Int): Long {
        return 0
    }

    override fun getView(i: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: LayoutInflater.from(mContext).inflate(R.layout.item_device, parent, false)
        val isCurrentDevice = mDevices[i].key.contentEquals(mCurrentDeviceId)
        val devId: TwoButtonEditText = view.findViewById(R.id.txt_device_id)
        val thisDevice = view.findViewById<TextView>(R.id.txt_device_thisflag)
        devId.text = mDevices[i].value
        var hint = mDevices[i].key
        hint = hint.substring(0, (hint.length * 0.66).toInt())
        devId.setHint(hint)
        if (isCurrentDevice) {
            thisDevice.visibility = View.VISIBLE
            devId.setLeftDrawable(R.drawable.baseline_edit_twoton_24dp)
            devId.setLeftDrawableOnClickListener { mListener.onDeviceRename() }
        } else {
            thisDevice.visibility = View.GONE
            devId.setLeftDrawable(R.drawable.baseline_cancel_24)
            devId.setLeftDrawableOnClickListener { mListener.onDeviceRevocationAsked(mDevices[i].key) }
        }
        return view
    }

    interface DeviceRevocationListener {
        fun onDeviceRevocationAsked(deviceId: String)
        fun onDeviceRename()
    }

    init {
        setData(devices, currentDeviceId)
    }
}